Nevergrad: OnePlusOne Optimiser addition

gulshan-123 opened the pull request:
I have written a basic structure of OnePlusOne optimisation wrapper.

gulshan-123 commented:
I have some doubts:
 -  When I try to print the lower and upper bounds, if no bounds are provided, it prints an array of size 10 filled with -10, regardless of the size of X0. This is causing an error in the set_bound function of Nevergrad.
 - The OnePlusOne function is designed to run for the specified budget (a parameter in Nevergrad) number of iterations, irrespective of the function’s values (no by default stopping condition, budget is also required parameter).
 - However, it has an option for early stopping using a boolean flag. How can I configure it so that both MAX_ITER and MAX_EVALUATION are set, while still allowing it to stop early if needed?

janosg commented:
> * When I try to print the lower and upper bounds, if no bounds are provided, it prints an array of size 10 filled with -10, regardless of the size of X0. This is causing an error in the set_bound function of Nevergrad.

This does not contain enough information to understand your problem. We need at least the following:
- What did you do
- What did you expect to happen
- What happened

Ideally you follow [this blogpost](https://matthewrocklin.com/minimal-bug-reports.html) when describing problems you encounter. 

> * The OnePlusOne function is designed to run for the specified budget (a parameter in Nevergrad) number of iterations, irrespective of the function’s values (no by default stopping condition, budget is also required parameter).

What is your question or doubt? If I understand correctly, this is how most global optimizers behave. 

> * However, it has an option for early stopping using a boolean flag. How can I configure it so that both MAX_ITER and MAX_EVALUATION are set, while still allowing it to stop early if needed?

Algorithms in optimagic can have as many options as you need and they can have any type. So it should not be a problem to allow complete configurability of the algorithm.

gulshan-123 commented:
Hi @janosg, I've corrected the error from my previous message. I've now added the OnePlusOne wrapper, and it's ready for your review.  There's a setting to control the maximum optimization time in Nevergrad – any ideas for a more descriptive name?  I'm still working on adding constraints.

janosg commented:
Hi @gulshan-123, thanks for the PR. I'll do a thorough review once your wrapper is feature complete. So here are just some quick comments to help you get there:

- The [OnePlusOne algorithm](https://facebookresearch.github.io/nevergrad/optimizers_ref.html#nevergrad.families.ParametrizedOnePlusOne) seems to support many more tuning parameters than what you currently expose in the wrapper. Examples are `noise_handling`, `mutation` and `crossover`. In optimagic we always want complete wrappers that allow full customization of the wrapped algorithms. 
- You set `is_global=False`; What is your source for that? I am not familiar with the algorithm but from everything I see, it looks like this is a genetic algorithm and those are usually global
- We need documentation for the algorithm, including sources like the paper that introduced the algorithm. 
- You currently have `disable_history=False` but you rely on the algorithm's parallelization. This will not work and there are two options around it:
    1. Using disable_history and find another way to collect the optimizer history. This is usually hard and not desirable for maintenance reasons
    2. Use `problem.batch_fun` for the parallelization; This should be possible if you don't use their `minimize` interface but the lower level [ask-and-tell interface](https://facebookresearch.github.io/nevergrad/optimizers_ref.html#nevergrad.optimizers.base.Optimizer.ask). You would use `ask` multiple times to create a batch of candidate parameters of size `n_cores` and then use `problem.batch_fun` to evaluate the function on this batch in parallel while preserving the history collection.
    
The documentation is an essential part of the PR and needs to convince us that you did a thorough job in exploring all the tuning parameters of the algorithm.

gulshan-123 commented:
> * You set `is_global=False`; What is your source for that? I am not familiar with the algorithm but from everything I see, it looks like this is a genetic algorithm and those are usually global

OnePlusOne, as implemented in Nevergrad, is a variant of the (1+1)-Evolution Strategy. It operates by maintaining a single current solution `a` and generating a mutated candidate `a + da`. The algorithm accepts the new point only if it improves the objective value. There is no population, no crossover, and no explicit exploration mechanism. Thus it should be local.
I also found it here: [Rastrigin function using 1+1-trapped in local optimum](https://youtu.be/3LQSa_ZhN0I) from the Medium Article: [EE-(1+1)](https://mgijon94.medium.com/the-starting-point-of-evolution-strategies-the-ee-1-1-algorithm-511ea3f2edf7)



> 2\. Use `problem.batch_fun` for the parallelization; This should be possible if you don't use their `minimize` interface but the lower level [ask-and-tell interface](https://facebookresearch.github.io/nevergrad/optimizers_ref.html#nevergrad.optimizers.base.Optimizer.ask). You would use `ask` multiple times to create a batch of candidate parameters of size `n_cores` and then use `problem.batch_fun` to evaluate the function on this batch in parallel while preserving the history collection.

I will try this.

gulshan-123 commented:
Hi @janosg , I have added all the parameters from nevergrad OnePlusOne in optimagic. 
I need to now add support for nonlinear-constraint.

janosg commented on src/optimagic/optimizers/nevergrad_optimizers.py:
```diff
@@ -0,0 +1,139 @@
+                "Nevergrad is not installed. Please install it to use this algorithm."
+            )
+        param = ng.p.Array(
+            init=np.clip(x0, problem.bounds.lower, problem.bounds.upper)
```
Why did you decide to clip the start parameters? Optimagic ensures that x0 is within bounds. Did you have a problem when not doing the clipping?

janosg commented on src/optimagic/optimizers/nevergrad_optimizers.py:
```diff
@@ -0,0 +1,139 @@
+        recommendation = optimizer.provide_recommendation()
+
+        result = InternalOptimizeResult(
+            x=recommendation.value[0][0],
+            fun=recommendation.loss,
+            success=True,
+            n_fun_evals=optimizer.num_ask,
```
Please set as many fields as possible for the `InternalOptimizeResult`, even if they are trivial. For example, `n_jac_evals` could be set to zero instead of None (which happens if you do nothing).

janosg commented on tests/optimagic/optimization/test_nevegrad.py:
```diff
@@ -0,0 +1,22 @@
+import optimagic as om
+
+
+def test_x0_works_in_minimize():
+    res = om.minimize(
+        fun=lambda x: x @ x,
+        x0=np.arange(3),
+        algorithm="nevergrad_oneplusone",
+    )
+    aaae(res.params, np.zeros(3))
+
+
+def test_x0_works_in_maximize():
+    res = om.maximize(
+        fun=lambda x: -x @ x,
+        x0=np.arange(3),
+        algorithm="nevergrad_oneplusone",
+    )
+    aaae(res.params, np.zeros(3))
```
Both of these tests are essentially duplicates of tests we already run for all optimizers (e.g. [here](https://github.com/optimagic-dev/optimagic/blob/main/tests/optimagic/optimization/test_many_algorithms.py). For algorithm specific test it would be better to write tests that show that the algorithm specific parameters have their intended effects. See [here](https://github.com/optimagic-dev/optimagic/blob/main/tests/optimagic/optimizers/test_tao_optimizers.py) for examples. 

If you don't find ways to test algorithm specific behavior, you can simply delete the test file and rely on our standard tests.

janosg reviewed (changes requested):
Hi @gulshan-123. Sorry for the late review. 

Besides the specific comments we need the following changes:

1. Install nevergrad via pip instead of conda to fix the test failures (the conda version is outdated)
2. Solve the merge conflicts with the current main branch (In the meantime, we have merged the first PR that adds another nevergrad optimizer, so there are a few conflicts).

gulshan-123 commented:
I first pulled the latest main branch and then made changes on that only to avoid any merge conflict later. 

The following test is failing.
```python
____________________________________________________________ test_algorithm_on_sum_of_squares_with_binding_bounds[nevergrad_oneplusone] _____________________________________________________________

algorithm = 'nevergrad_oneplusone'

    @pytest.mark.parametrize("algorithm", BOUNDED_ALGORITHMS)
    def test_algorithm_on_sum_of_squares_with_binding_bounds(algorithm):
        res = minimize(
            fun=sos,
            params=np.array([3, 2, -3]),
            bounds=Bounds(
                lower=np.array([1, -np.inf, -np.inf]), upper=np.array([np.inf, np.inf, -1])
            ),
            algorithm=algorithm,
            collect_history=True,
            skip_checks=True,
        )
        assert res.success in [True, None]
        decimal = 3
>       aaae(res.params, np.array([1, 0, -1]), decimal=decimal)
E       AssertionError: 
E       Arrays are not almost equal to 3 decimals
E       
E       Mismatched elements: 1 / 3 (33.3%)
E       Max absolute difference among violations: 0.32302791
E       Max relative difference among violations: inf
E        ACTUAL: array([ 1.   ,  0.323, -1.   ])
E        DESIRED: array([ 1,  0, -1])

```

As discussed here https://github.com/optimagic-dev/optimagic/pull/579#issuecomment-2835793667, I tried to change the default parameters, but was still failing. Even after increasing the budget, `res` was going closer to 0 (eg: 0.01), but still not passing the test.

gauravmanmode commented on src/optimagic/optimizers/nevergrad_optimizers.py:
```diff
@@ -108,3 +108,129 @@ def _solve_internal_problem(
+    name="nevergrad_oneplusone",
+    solver_type=AggregationLevel.SCALAR,
+    is_available=IS_NEVERGRAD_INSTALLED,
+    is_global=False,
```
Hi @gulshan-123, just a suggestion. but I think `is_global = True` would be more in line with the description given [here](https://pdf.sciencedirectassets.com/271538/1-s2.0-S0304397500X0415X/1-s2.0-S0304397501001827/main.pdf?X-Amz-Security-Token=IQoJb3JpZ2luX2VjEDgaCXVzLWVhc3QtMSJHMEUCID4%2BMZEE2aDxE9ZKC9TdJjx2eRi0fhvuuuJY6lNxTzPwAiEAi1fG8z4cPLSW0IAfmzLqRwkxLnYqlh%2FahL9iulgXeO4qvAUI8f%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FARAFGgwwNTkwMDM1NDY4NjUiDEOhotWEMBcP6vxXcSqQBU7MKqFHXKJZvWtZ9r9vjkH%2BXoYs%2FZrHRpOiuXm9t3%2BosaDv2FbCcaLp5PNamSyMDjLyyi%2FTIE%2Bdn4KPn5QazPNkcL5wcGGXJYmEsP4Lq5bA%2FLmtQ0oekefPZlqwvG%2FI9T0lQ71%2F0yWu0HluXksfQhZ6PjuOwDByew1z50O1L8PWEuArPw8l7%2FMTDNEBvBhSfo13yu%2ByXo18bUcV0iEmuh1rd%2FTfAbS20IEGE85LbQHwd0B1IKHq%2B8oSCzfX1ew2qr4Emlvb%2FIZzeudvN5pBRNfuLbKplX3nhCMHeThWvlyjJm3GNN8E%2BWw7A2k2Elggyx1NeLk0G%2BS%2F2t5q9j9KuOWdMGK4L6I5XntaIXWk5BaorBszrXhXnt6Zo2Ce10Qg3LNWhFmO%2BtaofsfGFEcYDlBa2g%2BsFdNnfF21C8a9558pxlM3%2FUl5sV1s5Ojzn4bXkkRx6MRQJTlMhD%2FETzN8tABwPUZ7yG1nhCSUCtevGgLbkkuzxPLtuPbSh%2BdyDGvczoQQDpVACjAp4W%2FXuNyOZvJDJq3SHo%2BG5I1qonFaf3gj2hXnz6ZM8ALVh0n8QC6QJ%2FBrYIFllc69HkYpe%2B9%2Bnl30cUyvI%2FUQ%2F4tagFyD3GqdxxXCQ5sWUE6rkfI2n2Q0%2ByM6QbY4S1XPizOfdwz%2Be1ChDyy4kSRbqze8KW%2B4xdcjPs5SbdGKHfJN7JDkbpCFSUaoG6%2BA9n8qnQasSrnRbswG%2FvDoqq%2B8JHjJdzkwCrFv0upCH3VAJO0fOTCRd%2FKDP3Ou8Ko8otVbCyOZxx4DOMv8FEzHvMRQgSC6KKsuAoob0vxkzvQHcxKfkJWiaK9Q6Iz5CgT7KMMIKMdZrWl2NQHyt04MFUXJsZ1wj7SvxwkpMM6pwsEGOrEB7ne%2Fprsx3BB6ZuvDYSEVvmHGNdQJAsk6I6PhoNUM6DjFbZWQH2ah95OsTHi%2FYWIQXcVsm1J8adpmBg232RHP9%2FtgplBsxjMwSbZ%2BP6eCzmkHZlxwa0b%2BJliZGt5X0IO%2Bq%2BbkOAlPHMTKF6VHLeWwn8bIgIIfnKD3M8XuLZvld3Cf8MzcsQATC9L2F5iwB7%2FauZqB05I92MjTuzodZxRd9MydFCu5kKuZiO8ePwGIC8TU&X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Date=20250523T161229Z&X-Amz-SignedHeaders=host&X-Amz-Expires=300&X-Amz-Credential=ASIAQ3PHCVTY7HHLORSK%2F20250523%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Signature=0d156752257e1c24d0ac1920d3f78763ad3c87a398f41fb00b9766b6bfb9b057&hash=f9c0347384fb9710db2ab9d921c1c5f6300ab5caf9013e3841ee3ed918f98b91&host=68042c943591013ac2b2430a89b270f6af2c76d8dfd086a07176afe7c76c2c61&pii=S0304397501001827&tid=spdf-49e55353-826c-439e-9884-84ea4f01153f&sid=9fae316f239a7440538884e273c3d41bb230gxrqb&type=client&tsoh=d3d3LnNjaWVuY2VkaXJlY3QuY29t&rh=d3d3LnNjaWVuY2VkaXJlY3QuY29t&ua=13005e5a525100530259&rr=9445d5c8bdccc179&cc=in).

gulshan-123 replied:
Okay, Also, please tell what changes to do to pass all the CI. It is failing because the answer is not converging exactly to 0. something close to 0 is resulting.

gauravmanmode replied:
The tests should pass if you set `is_global=True`. The test`test_stochastic_algorithm_on_sum_of_squares_with_binding_bounds` was failing because it is not a  `LOCAL_STOCHASTIC_ALGORITHM`. 
Also, it would be nice if u can add docs like description of the algorithm and parameters names.

gulshan-123 replied:
Yes, it is passing after setting `is_global=True` Thank You. I will now add some docs and references.

gulshan-123 commented:
Hi @gauravmanmode, i have added the docs and all the pytests are passing. Thank you very much.

janosg commented on docs/source/algorithms.md:
```diff
@@ -4043,6 +4043,87 @@ these optimizers, you need to have
+
+    Minimize a scalar function using the One Plus One algorithm.
+
+    The Particle Swarm Optimization algorithm was originally proposed by
+    :cite:`Rechenberg1973`. The implementation in Nevergrad is based on
+    the one-fifth adaptation rule, going back to :cite:`Schumer1968`, discovered independently by :cite:`devroye1972` and :cite:`Rechenberg1973`.
```
Can you describe better how OnePlusOne relates to PSO?

janosg commented on docs/source/algorithms.md:
```diff
@@ -4043,6 +4043,87 @@ these optimizers, you need to have
+    perhaps more aggressively. If the parent yields a better result, it 
+    remains the parent and the next perturbation is less aggressive.
+
+    - **noise\_handling** (str or Tuple\[str, float] or None): Method for handling the noise. The name can be (Default: `None`).
+      - `"random"`: A random point is reevaluated regularly using the one-fifth adaptation rule.
+      - `"optimistic"`: The best optimistic point is reevaluated regularly, embracing optimism in the face of uncertainty.
+      - A float coefficient can be provided to tune the regularity of these reevaluations (default is 0.05).
```
We need a bit more intuition. What is done with the result of the re-evaluation and how does it help in a noisy setting? What would a value of 0.05 mean? Is the re-evaluation done in 5 % of the iterations?

janosg reviewed (commented):
Hi @gulshan-123, the code looks good. I made a few comments in the documentation. We can merge after they are addressed.

gulshan-123 commented:
@janosg, I have changed as suggested. Tests are still failing due to answer coming close to 0 instead of exact 0.

janosg commented:
@gulshan-123 the tests are passing now. The key was to mark the algorithm as global. The precision requirements for global algorithms are lower than for local ones.

gulshan-123 commented on src/optimagic/optimizers/nevergrad_optimizers.py:
```diff
@@ -108,3 +108,129 @@ def _solve_internal_problem(
+    annealing: Literal[
+        "none", "Exp0.9", "Exp0.99", "Exp0.9Auto", "Lin100.0", "Lin1.0", "LinAuto"
+    ] = "none"
+    sparse: bool | int = False
```
In nevergrad, this is Union[bool, int], so I decided to use that only. [L122](https://github.com/facebookresearch/nevergrad/blob/main/nevergrad/optimization/optimizerlib.py#L122). This is used as int here: [L429-434](https://github.com/facebookresearch/nevergrad/blob/main/nevergrad/optimization/optimizerlib.py#L429). Should I use only bool?

janosg replied:
If the int is just assessed for truthyness I would only support a bool. If the int enables to pass information that could not be passed via a bool (e.g. how sparse it is) we need to support and document it.

gulshan-123 commented on docs/source/algorithms.md:
```diff
@@ -4043,6 +4043,87 @@ these optimizers, you need to have
+    perhaps more aggressively. If the parent yields a better result, it 
+    remains the parent and the next perturbation is less aggressive.
+
+    - **noise\_handling** (str or Tuple\[str, float] or None): Method for handling the noise. The name can be (Default: `None`).
```
In this case, nevergrad will assume that the function is deterministic.

gulshan-123 commented on src/optimagic/optimizers/nevergrad_optimizers.py:
```diff
@@ -108,3 +108,129 @@ def _solve_internal_problem(
+    crossover_type: Literal["none", "rand", "max", "min", "onepoint", "twopoint"] = (
+        "none"
+    )
+    tabu_length: int = 1000
```
It is in fact NonNegativeInt: https://github.com/facebookresearch/nevergrad/blob/b2fec24d93a1f79e40880b462887a0c8be0e1a25/nevergrad/optimization/optimizerlib.py#L582

gauravmanmode commented on src/optimagic/optimizers/nevergrad_optimizers.py:
```diff
@@ -108,3 +108,129 @@ def _solve_internal_problem(
+)
+@dataclass(frozen=True)
+class NevergradOnePlusOne(Algorithm):
+    noise_handling: str | Tuple[str, float] | None = None
```
Here, you can be more specific like
 `noise_handling: Literal["random", "optimistic"] | Tuple[Literal["random", "optimistic"], float] | None = None`

gauravmanmode commented on docs/source/algorithms.md:
```diff
@@ -4043,6 +4043,88 @@ these optimizers, you need to have
+    - **noise\_handling** (str or Tuple\[str, float] or None): Method for handling the noise. The name can be (Default: `None`).
+      - `"random"`: A random point is reevaluated regularly using the one-fifth adaptation rule.
+      - `"optimistic"`: The best optimistic point is reevaluated regularly, embracing optimism in the face of uncertainty.
+      - `"none"`: In this case, the algorithm assumes the function is deterministic.
```
You cannot pass `"none"` as a option when the default is `None`

gauravmanmode commented on src/optimagic/optimizers/nevergrad_optimizers.py:
```diff
@@ -108,3 +108,129 @@ def _solve_internal_problem(
+                " for more detailed installation instructions."
+            )
+
+        instrum = ng.p.Array(
```
to avoid inf values when bounds are not provided, set bounds through` instrum = ng.p.Array(init=x0).set_bounds(
            lower=problem.bounds.lower, upper=problem.bounds.upper )`

gauravmanmode commented:
Hi @janosg, I have fixed some things and moved the documentation to docstring.
 Can you review this, suggest any changes and I can take this over in the other PR?

janosg reviewed (approved):
@gauravmanmode, thanks for taking over! A few tiny comments but I already approve. You can fix the comments and then merge without another review.

gauravmanmode commented on src/optimagic/optimizers/nevergrad_optimizers.py:
```diff
@@ -108,3 +108,221 @@ def _solve_internal_problem(
+    roulette_size: PositiveInt = 64
+    antismooth: NonNegativeInt = 4
+    crossover: bool = False
+    crossover_type: Literal["none", "rand", "max", "min", "onepoint", "twopoint"] = (
```
nevergrad uses "none" as a str but we can handle both cases.

gauravmanmode commented:
I'm sorry I totally forgot about this. @janosg Should i resolve the conflicts and merge this?
